src/lib/UniversalDiscovery/CotfPermissionChecker.php

```diff
@@ -0,0 +1,305 @@
+                    }
+
+                    if (!empty($possibleRestrictedContentTypes)) {
+                        $restrictedContentTypesIds[] = $possibleRestrictedContentTypes;
+                    }
+                }
+            }
+        }
+
+        return new CotfCreateRestrictions(
+            true,
+            array_unique(array_merge(...$restrictedContentTypesIds)),
+            array_unique(array_merge(...$restrictedLanguagesCodes))
+        );
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\User\Limitation $limitation
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
+     * @param \eZ\Publish\SPI\Limitation\Type $type
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $currentUserRef
+     * @param \eZ\Publish\API\Repository\Values\Content\LocationCreateStruct $locationCreateStruct
+     *
+     * @return array
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    private function getPossibleLanguageCodes(
+        Limitation $limitation,
+        Location $location,
+        Type $type,
+        UserReference $currentUserRef,
+        LocationCreateStruct $locationCreateStruct
+    ): array {
+        $languageCodes = [];
+        foreach ($limitation->limitationValues as $languageLimitationValue) {
+            $contentCreateStruct = $this->createContentCreateStruct($location, new CoreContentType(), $languageLimitationValue);
+            $access = $type->evaluate($limitation, $currentUserRef, $contentCreateStruct, [$locationCreateStruct]);
+
+            if ($access === LimitationType::ACCESS_GRANTED) {
+                $languageCodes[] = $languageLimitationValue;
+            }
+        }
+
+        return $languageCodes;
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\User\Limitation $limitation
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
+     * @param \eZ\Publish\SPI\Limitation\Type $type
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $currentUserRef
+     * @param \eZ\Publish\API\Repository\Values\Content\LocationCreateStruct $locationCreateStruct
+     *
+     * @return array
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    private function getPossibleContentTypes(
+        Limitation $limitation,
+        Location $location,
+        Type $type,
+        UserReference $currentUserRef,
+        LocationCreateStruct $locationCreateStruct
+    ): array {
+        $contentTypeList = $this->contentTypeService->loadContentTypeList($limitation->limitationValues);
+        $possibleContentTypes = [];
+        foreach ($contentTypeList as $contentType) {
+            $contentCreateStruct = $this->createContentCreateStruct($location, $contentType, '');
```

Comment:
I don't think it's going to be a reliable check for custom limitations. Since it related to our CORE extensibility point you can easily break custom code. Moreover `CustomType` is an entity and you assume it's valid. This intentionally passes invalid `ContentType` and language code.